sdk/src/user.ts

```diff
@@ -116,6 +140,303 @@ export class User {
+		}
+	): UserMarginCalculation {
+		const strict = opts?.strict ?? false;
+		const enteringHighLeverage = opts?.enteringHighLeverage ?? false;
+		const includeOpenOrders = opts?.includeOpenOrders ?? true; // TODO: remove this ??
+		const marginBuffer = opts?.liquidationBuffer; // treat as MARGIN_BUFFER ratio if provided
+		const marginRatioOverride = opts?.marginRatioOverride;
+
+		// Equivalent to on-chain user_custom_margin_ratio
+		let userCustomMarginRatio =
+			marginCategory === 'Initial' ? this.getUserAccount().maxMarginRatio : 0;
+		if (marginRatioOverride !== undefined) {
+			userCustomMarginRatio = Math.max(
+				userCustomMarginRatio,
+				marginRatioOverride
+			);
+		}
+
+		// Initialize calc via JS mirror of Rust MarginCalculation
+		const ctx = MarginContext.standard(marginCategory)
+			.strictMode(strict)
+			.setMarginBuffer(marginBuffer)
+			.setMarginRatioOverride(userCustomMarginRatio);
+		const calc = new JsMarginCalculation(ctx);
+
+		// SPOT POSITIONS
+		for (const spotPosition of this.getUserAccount().spotPositions) {
+			if (isSpotPositionAvailable(spotPosition)) continue;
+
+			const spotMarket = this.driftClient.getSpotMarketAccount(
+				spotPosition.marketIndex
+			);
+			const oraclePriceData = this.getOracleDataForSpotMarket(
+				spotPosition.marketIndex
+			);
+			const twap5 = strict
+				? calculateLiveOracleTwap(
+						spotMarket.historicalOracleData,
+						oraclePriceData,
+						new BN(Math.floor(Date.now() / 1000)),
+						FIVE_MINUTE
+				  )
+				: undefined;
+			const strictOracle = new StrictOraclePrice(oraclePriceData.price, twap5);
+
+			if (spotPosition.marketIndex === QUOTE_SPOT_MARKET_INDEX) {
+				const tokenAmount = getSignedTokenAmount(
+					getTokenAmount(
+						spotPosition.scaledBalance,
+						spotMarket,
+						spotPosition.balanceType
+					),
+					spotPosition.balanceType
+				);
+				if (isVariant(spotPosition.balanceType, 'deposit')) {
+					// add deposit value to total collateral
+					const tokenValue = getStrictTokenValue(
+						tokenAmount,
+						spotMarket.decimals,
+						strictOracle
+					);
+					calc.addIsolatedTotalCollateral(tokenValue);
+				} else {
+					// borrow on quote contributes to margin requirement
+					const tokenValueAbs = getStrictTokenValue(
+						tokenAmount,
+						spotMarket.decimals,
+						strictOracle
+					).abs();
+					calc.addIsolatedMarginRequirement(
+						tokenValueAbs,
+						tokenValueAbs,
+						MarketIdentifier.spot(0)
+					);
+					calc.addSpotLiability();
+				}
+				continue;
+			}
+
+			// Non-quote spot: worst-case simulation
+			const {
+				tokenAmount: worstCaseTokenAmount,
+				ordersValue: worstCaseOrdersValue,
+				tokenValue: worstCaseTokenValue,
+				weightedTokenValue: worstCaseWeightedTokenValue,
+			} = getWorstCaseTokenAmounts(
+				spotPosition,
+				spotMarket,
+				strictOracle,
+				marginCategory,
+				userCustomMarginRatio
+			);
+
+			// open order IM
+			if (includeOpenOrders) {
```

Comment:
The includeOrders is a concept that only exist in sdk, we also included it for program margin calc. So this doesnt really make sense here if we trying to replicate onchain

Would be good to review how ui uses include open orders, should see if we can get rid of it

Reply:
I went back and forth buth unfortunately it seems like maybe we need this value. 
Because of code like this: https://github.com/drift-labs/drift-common/blob/master/common-ts/src/common-ui-utils/trading.ts#L199

That is passing `undefined` but `liquidationPrice()` on user defaults to false for this value. So in some liquidation price cases we want to exclude open orders from the calculation. 

Not sure if that means we should not use this new getMarginCalculation thing for liquidation prices but idk if that makes sense. Or it means we need to just keep the functionality. Open to your thoughts.

Reply:
I see, yeah i guess we can keep it then